MAINT: stats: remove `mvn` fortran calls from `multivariate_normal.cdf`

ev-br opened the pull request:
#### Reference issue
<!--Example: Closes gh-WXYZ.-->

https://github.com/scipy/scipy/pull/17410

closes https://github.com/scipy/scipy/issues/8367, closes https://github.com/scipy/scipy/issues/16142

#### What does this implement/fix?
<!--Please explain your changes.-->

Use the python translation of the multivariate_normal integrals instead of the Fortran `_mvn` extension. IIUC the python integration code is a translation from matlab, where both matlab and fortran code are from the same author, Alan Genz,
cf a discussion at https://github.com/scipy/scipy/pull/17410, esp https://github.com/scipy/scipy/pull/17410#issuecomment-1312628547

#### Additional information
<!--Any additional information you think is important.-->

ATM there are at least two issues:

1. the replacement code is obscenely slow. Probably needs some cythonizing or pythranizing, as discussed in gh-17410 and elsewhere
2. There are numerical issues in some of our tests, had to apply a rather ugly workaround to avoid propagating nans.

ev-br commented:
The python translation , https://github.com/scipy/scipy/blob/main/scipy/stats/_qmvnt.py#L428, does what looks like a Cholesky factorization of the covariance matrix _plus_ some handling of the integration bounds on top. The procedure mixes the factorization and transforming the integration bounds in a bespoke double loop. Am wondering if there's a reasonable way to disentangle the two and use a standard LAPACK call for the Cholesky factorization. Am wondering if you've tried something like this when translating the code @rkern?

mdhaber commented on scipy/stats/_qmvnt.py:
```diff
@@ -251,7 +251,13 @@ def _qmvn(m, covar, low, high, rng, lattice='cbc', n_batches=10):
             x = abs(2 * z - 1)
-            y[i - 1, :] = phinv(c + x * dc)
+
+            # work around inf - inf = nan in the matmul below
```
In what cases does it come up?

ev-br replied:
```
$ python dev.py test -t scipy/stats/tests/test_multivariate.py::TestMultivariateT::test_cdf_against_multivariate_normal -- --pdb

....

>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB set_trace (IO-capturing turned off) >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> /home/br/repos/scipy/scipy/build-install/lib/python3.12/site-packages/scipy/stats/_qmvnt.py(265)_qmvn()

.....

(Pdb) p cho
array([[ 1.        ,  0.        ,  0.        ,  0.        ,  0.        ],
       [ 1.95700502,  1.        ,  0.        ,  0.        ,  0.        ],
       [ 1.65026966,  0.03347696,  1.        ,  0.        ,  0.        ],
       [ 2.00512128,  0.54881632,  0.41484905,  1.        ,  0.        ],
       [ 2.74194625,  0.07662511, -0.15581473,  0.87468593,  1.        ]])
(Pdb) p np.isnan(s).any()
np.True_
(Pdb) p np.isnan(y).any()
np.False_
(Pdb) p np.isnan(cho[i, :i] @ y[:i, :]).any()
*** RuntimeWarning: invalid value encountered in matmul

(Pdb) np.argwhere(y[:i, :] == np.inf)
array([[     1, 257691],
       [     1, 257692],
       [     1, 257693]])
(Pdb) np.argwhere(y[:i, :] == -np.inf)
array([[     2, 257691],
       [     2, 257692],
       [     2, 257693]])
(Pdb) p y[:i, 257690:257695]
array([[-4.43818872, -4.58867698, -5.25179331, -4.57501357, -4.43115065],
       [ 8.20953615,         inf,         inf,         inf,  8.20953615],
       [ 0.54934329,        -inf,        -inf,        -inf,  2.15475497]])
```

rkern commented:
I considered it, but concluded that it wasn't worth the thinking effort:

- The same possibilities and tradeoffs existed in the MATLAB code, where he could have used the off-the-shelf Cholesky and do the transformation on the side, but didn't. Which suggests to me that the independent transformation code might take just as long as doing both at the same time.
- Pythranization will likely optimize it into nothingness for most problems.
- It's not in the hot loop.

ev-br commented:
Thanks @rkern! I suspected as much, great to have a confirmation from the horse's mouth. 

--------------------------

The current iteration offloads inner loops of both `qmvt` and `qmvn` to pythran. Interestingly,
- the result is *very* slow
- looking at the profile, it seems that the bottleneck is computing the error function (a profile's below the fold)
- pythran transpiles `scipy.special.ndtr` into `boost::math::erf`, unlike scipy.special, which IIUC uses a version from cephes.



<details>

```
$ cat prof_mvn.py
import numpy as np
from scipy.stats import multivariate_normal

n = 4

data = np.random.randn(n, n)
cov = np.dot(data, data.T)
mean = np.random.randn(n)

# Construct an ndarray which can be interpreted as
# a 2x3 array whose elements are random data vectors.
X = np.random.randn(2, 3, n)

desired_cdf = multivariate_normal.cdf(X, mean, cov)
print(desired_cdf)

-----------------------------------------

$ python -m yep --visualize -- prof_mvn.py

Total: 1265 samples
       0   0.0%   0.0%     1225  96.8% PyEval_EvalCode
       9   0.7%   0.7%     1225  96.8% _PyEval_EvalFrameDefault.cold
       0   0.0%   0.7%     1225  96.8% builtin_exec
       0   0.0%   0.7%     1225  96.8% cfunction_vectorcall_FASTCALL_KEYWORDS
       0   0.0%   0.7%     1218  96.3% PyObject_Vectorcall
       0   0.0%   0.7%     1210  95.7% Py_RunMain.localalias
       0   0.0%   0.7%     1210  95.7% pymain_run_module
       0   0.0%   0.7%     1208  95.5% Py_BytesMain
       0   0.0%   0.7%     1205  95.3% __libc_start_main
       0   0.0%   0.7%     1203  95.1% _start
       0   0.0%   0.7%     1195  94.5% dispatcher_vectorcall
       0   0.0%   0.7%     1178  93.1% _PyObject_MakeTpCall.localalias
       0   0.0%   0.7%     1161  91.8% cfunction_call
      17   1.3%   2.1%     1124  88.9% __pythran_wrap__qmvn_inner0 [clone .constprop.0]
       0   0.0%   2.1%     1124  88.9% __pythran_wrapall__qmvn_inner
      67   5.3%   7.4%      735  58.1% ::pythonic::types::ndarray::ndarray
      57   4.5%  11.9%      668  52.8% std::enable_if::type ::pythonic::scipy::special::functor::ndtr::operator [clone .constprop.0] [clone .isra.0]
     197  15.6%  27.4%      611  48.3% boost::math::detail::erf_imp [clone .isra.0]
      35   2.8%  30.2%      361  28.5% expf64x
       8   0.6%  30.8%      339  26.8% ::pythonic::types::ndarray::ndarray [clone .constprop.0]
      64   5.1%  35.9%      331  26.2% std::enable_if::type ::pythonic::scipy::special::functor::ndtri::operator [clone .constprop.0] [clone .isra.0]
     326  25.8%  61.7%      326  25.8% __expl_finite
     200  15.8%  77.5%      267  21.1% boost::math::detail::erf_inv_imp [clone .isra.0]
      57   4.5%  82.0%       57   4.5% __lgammal_r_finite
      28   2.2%  84.2%       42   3.3% ldexpf64x
      40   3.2%  87.4%       40   3.2% blas_thread_server
```
</details>

serge-sans-paille commented:
@ev-br : 
it's a pythran limitation (performance wise), but does that change improve the performance?

```
                for jp in range(s.shape[1]):
                    s[i:,jp] += cho[i:, i - 1] * y
```

instead of

```
                s[i:, :] += cho[i:, i - 1][:, np.newaxis] * y
```

ev-br commented:
TBH, I did not look at the perf of _qmvt, not yet anyway. The elephant in the room is the 25x slowdown in qmvn, and the profile is dominated by the error function and it's inverse. The xsf/cephes version is literally 50 lines of C to evaluate a rational approximation, so I wanted to check how hard would it be to make pythran use that instead of boost or what it transpiles to for scipy.special.ndtr.

serge-sans-paille commented:
On Sun, Jan 12, 2025 at 12:14:38PM -0800, Evgeni Burovski wrote:
> TBH, I did not look at the perf of _qmvt, not yet anyway. The elephant in the
> room is the 25x slowdown in qmvn, and the profile is dominated by the error
> function and it's inverse. The xsf/cephes version is literally 50 lines of C to
> evaluate a rational approximation, so I wanted to check how hard would it be to
> make pythran use that instead of boost or what it transpiles to for
> scipy.special.ndtr.

ok, I'll have a look at this on my side too and report here.

serge-sans-paille commented:
I don't think `ndtr` is the issue there:

```
(pythran) ssp@lakota:~/sources/pythran (master)% python -m timeit -s 'from scipy.special import ndtr' 'ndtr(0.5)'
1000000 loops, best of 5: 323 nsec per loop
(pythran) ssp@lakota:~/sources/pythran (master)% cat ndtr.py 
#pythran export ndtr(float64)
def ndtr(x):
    from scipy import special
    return special.ndtr(x)
(pythran) ssp@lakota:~/sources/pythran (master)% python -m pythran.run ndtr.py                                   
(pythran) ssp@lakota:~/sources/pythran (master)% python -m timeit -s 'from ndtr import ndtr' 'ndtr(0.5)'         
2000000 loops, best of 5: 142 nsec per loop
```

ev-br commented:
Tried expicit `#pythran export` of ndtr and ndtri, but this does not seem to change the profile: what took 0.5 sec with `_mvn.mvnun` now takes 13 sec, and 


```
Total: 1412 samples
       0   0.0%   0.0%     1370  97.0% PyEval_EvalCode
       9   0.6%   0.6%     1370  97.0% _PyEval_EvalFrameDefault.cold
       0   0.0%   0.6%     1370  97.0% builtin_exec
       0   0.0%   0.6%     1370  97.0% cfunction_vectorcall_FASTCALL_KEYWORDS
       1   0.1%   0.7%     1361  96.4% PyObject_Vectorcall
       0   0.0%   0.7%     1352  95.8% pymain_run_module
       0   0.0%   0.7%     1351  95.7% Py_RunMain.localalias
       0   0.0%   0.7%     1350  95.6% Py_BytesMain
       0   0.0%   0.7%     1348  95.5% __libc_start_main
       0   0.0%   0.7%     1346  95.3% _start
       0   0.0%   0.7%     1337  94.7% dispatcher_vectorcall
       0   0.0%   0.7%     1319  93.4% _PyObject_MakeTpCall.localalias
       0   0.0%   0.7%     1306  92.5% cfunction_call
      19   1.3%   2.1%     1268  89.8% __pythran_wrap__qmvn_inner0 [clone .constprop.0]
       0   0.0%   2.1%     1268  89.8% __pythran_wrapall__qmvn_inner
      81   5.7%   7.8%     1197  84.8% ::pythonic::types::ndarray::ndarray
      41   2.9%  10.7%      800  56.7% std::enable_if::type ::pythonic::scipy::special::functor::ndtr::operator [clone .constprop.0] [clone .isra.0]
     225  15.9%  26.6%      759  53.8% boost::math::detail::erf_imp [clone .isra.0]
      51   3.6%  30.2%      474  33.6% expf64x
     423  30.0%  60.2%      423  30.0% __expl_finite
      62   4.4%  64.6%      316  22.4% std::enable_if::type ::pythonic::scipy::special::functor::ndtri::operator [clone .constprop.0] [clone .isra.0]
     202  14.3%  78.9%      254  18.0% boost::math::detail::erf_inv_imp [clone .isra.0]
      46   3.3%  82.2%       46   3.3% __lgammal_r_finite
      29   2.1%  84.2%       43   3.0% ldexpf64x
      42   3.0%  87.2%       42   3.0% blas_thread_server
```

ev-br committed:
MAINT: stats: cythonize _qmvn_inner loop

The result seems to be 2x faster than the pythranized version.

ev-br commented:
Moving from pythran to cython and using `ndtr/ndtri` kernels from `cython_special` improves performance by about a factor of 2. (cf https://github.com/ev-br/scipy/commit/932300637a69949051a457699bc1625b8ddc9652). 

However, an interesting question is the number of evaluations for QMC integration. The default is `maxpts=1000000*dim`, and I wonder where does this factor comes from. 
The last commit decreases the default to `maxpts=10000*dim`, and tests pass withing a reasonable runtime. 

So the relevant questions are 1) where we control the accuracy anyhow, and 2) do we have some sort of accuracy target.
A target of this sort, however rough, is I'd say a prerequisite for any kind of further performance tuning.

ev-br committed:
TST: stats: add tests against semi-analytic MVN cases

For special structure covariance matrices, CDF can be reduced
to a univariate integral. The code and references are cf
`test_multivariate.py::MVNProblem`

rkern commented:
The nice thing about this new implementation is that it comes with its own estimate of its error, which has been fairly reliable in my investigations. Take a look at `_qauto()` for an example of using it to automatically add points until you reach a desired error bound (albeit at MC convergence rates rather than QMC convergence rates). `1_000_000 * ndim` is very, very high. `10_000 * ndim` seems to be a good rule of thumb, but YMMV depending on what precision you want to achieve in this context. For the use case of integrating `gaussian_kde`s, you probably don't need much because the errors for each Gaussian will likely cancel each other out.

See [this gist](https://gist.github.com/rkern/8518cb3e3f57fa81d21dbd9eabcfaa87#file-mvtestproblems-ipynb) for a widened family of test problems whose precise answers can be computed numerically with double integration.

rkern committed:
TST: stats: add MVNProblem and SingularMVNProblem

josef-pkt commented:
Doesn't maxpts only matter if releps and abseps are not satisfied?

cdf in the mutlivariate distribution does not allow for convergence options. So maxpts was set high when the cdf was added to the distribution.

(I had problems figuring out robust options for the convergence that can be taken as given by calling code, a long time ago when I wanted to add multivariate normal and t cdf to statsmodels. I didn't manage it and never used it for public features.)

rkern commented:
Yes, the old code `mvnun` does something similar to what `_qauto()` does; it iteratively adds more sample points until its estimate of the error meets the `abserr` and `relerr` criteria or if it hits `maxpts` early. So `maxpts` in the old code is not the same thing as `m` in `_qmvn()`, but rather `limit` in `_qauto()`.

ev-br committed:
MAINT: stats: remove `mvnun` from multivariate_normal

ev-br committed:
MAINT: stats: use an adaptive procedure in mvn.cdf

rkern commented on scipy/stats/_multivariate.py:
```diff
@@ -629,24 +632,27 @@ def _cdf(self, x, mean, cov, maxpts, abseps, releps, lower_limit):
-                return _mvn.mvnun(limits[:n], limits[n:], mean, cov,
-                                maxpts, abseps, releps)[0]
+            # res0 = _qmvn(maxpts, cov, limits[:n], limits[n:], rng)[0]
+            res = _qauto(_qmvn, cov, limits[:n], limits[n:],
```
If it's desired, then `_qauto` could straightforwardly be modified to consider both `abseps` and `releps` to match the old behavior.

ev-br replied:
Totally, it's straightforward. It's an "if" indeed though. I don't have a good feeling whether the full generality of `abseps` and `releps` is ever useful TBH.

rkern replied:
Probably not in this case. The main uses are where the box limits encompass the main body of the distribution, so the target value is within an order of magnitude near unity.

ev-br commented:
In an update to this PR, I added several tests against covariance matrices of special structure, cf https://github.com/scipy/scipy/pull/22298#issuecomment-2610344338
Please let me know @rkern  if you prefer not be listed as an author of https://github.com/scipy/scipy/pull/22298/commits/388917d69dbeb24973d9e3c71f85f71a127faca2

Curiously, adding these tests to _main_, where `multivariate_normal` delegates to
the original `mvnun` Fortran code, shows that original implementation is not without issues:
- the runtime is not predictable, rather slow in some cases
- the tolerance to which the semi-analytical tests pass varies from a test to test.
  While atol=1e-5 is mostly OK and often can be tightened further, is some cases
  (curiously, ndim=2, 3), it is as bad as 1e-1..1e-2.

So the original implementation is not a paragon of accuracy and performance. 
(Is obvious in retrospect, was not obvious a priori, not to me anyway).
My take from this is that we have a bit of leeway in balancing what the new implementation
improves with what is pessimizes. 

With that in mind, reusing an adaptive integrator, `_qauto`, shows a roughly similar performance and accuracy characteristics: some tolerances need small bumps, runtime is a little worse in some cases etc, but nothing drastic.
One curious thing is that semi-analytic tests with singular covariances crash across CI while running perfectly fine locally on a linux machine. (To check, revert https://github.com/scipy/scipy/pull/22298/commits/31a7d2de32488337b39f1f83f22026dcdab6d681)
Backtraces on CI were pointing at the pythranized calls of `_qmvn_inner`. Not sure what to make of these. 
We can probably leave these tests skipped or mark them `xslow`, they are brittle enough anyway.

rkern commented:
> * While atol=1e-5 is mostly OK and often can be tightened further, is some cases
>   (curiously, ndim=2, 3), it is as bad as 1e-1..1e-2.

Interesting. The `mvnun` implementation has a specialized code path for the bivariate case which should be rather more accurate. I'm wondering if my semi-analytic test case implementation is less robust in that case for some reason (if not outright bugged).

What was the final outcome of the Cythonization experiment? Was it worse in terms of accuracy or maintainability such that the performance increase did not pay for those pessimizations?

ev-br commented:
> What was the final outcome of the Cythonization experiment? Was it worse in terms of accuracy or maintainability such that the performance increase did not pay for those pessimizations?

There's no final outcome yet, it'll come later this coming week. 
One thing I'll try checking if it avoids pythran crashes on CI. if it does, it'll be a reasonably large argument in favor. 
Otherwise, a deal-breaker  is maintainability, I suppose --- which is in they eye of the current active scipy.stats maintainers.

mdhaber commented:
Most test failures seem to be gh-22408.
The other is a test timeout:
```python3
_________________ TestMultivariateNormal.test_cdf_vs_cubature _________________
[gw1] win32 -- Python 3.10.11 C:\hostedtoolcache\windows\Python\3.10.11\x64\python.exe
Test passed but took too long to run: Duration 56.580242299999554s > 5.0s
```

mdhaber commented:
OK, now failures are just gh-22395 and timeouts.

ev-br committed:
MAINT: fix _cy.pyx, make a slow test smaller

ev-br commented:
> What was the final outcome of the Cythonization experiment? Was it worse in terms of accuracy or maintainability such that the performance increase did not pay for those pessimizations?

The tip of this PR has a cythonized version of `_qmvn_inner`. The differences w.r.t. the pythranized version:
- the cythonized version is marginally faster, by a factor of ~1.5x - 2x
- the binary sizes of the cythonized and pythranized versions are about the same, ~200-250k on x86
- the cythonized version passes `_vs_univariate_singular` tests on CI, which were segfaulting with pythranized version.
(it's rather probable that the problem was my error, but I don't see where the error was)

All in all, my suggestion would be to use the cython version, unless there are strong opinions to the contrary. Thoughts?

rkern reviewed (commented):
Yeah, I think the Cython version is about as clear as the Pythran version. There's really just the one place where Pythran avoids unrolling the elementwise multiplication. I think continuing with Cythonizing `_qmvt_inner` is probably the last thing to tie this up.

Let me know if you want me to look more into the accuracy issues. Thank you for taking this on!

mdhaber commented on scipy/stats/tests/test_multivariate.py:
```diff
@@ -810,6 +1038,67 @@ def test_cdf_signs(self):
+        rng = np.random.default_rng(123)
+        a = rng.uniform(size=(ndim, ndim))
+        cov = a.T @ a
+        dist = multivariate_normal(np.zeros(ndim), cov=cov)
```
Could we do a nonzero mean in this test, at least?

ev-br replied:
Done.

mdhaber commented on scipy/stats/tests/test_multivariate.py:
```diff
@@ -810,6 +1038,67 @@ def test_cdf_signs(self):
+        npts = 1
+        x = rng.uniform(low=-3, high=3, size=(npts, ndim))
+        cdf = dist.cdf(x)
+        cdf_i = cubature(dist.pdf, [-np.inf]*ndim, x[j, ...]).estimate
```
`j` is not defined?

ev-br replied:
Thanks, fixed.

mdhaber reviewed (commented):
I did my own investigation into the performance and accuracy.

- I'm quite confident that we're not losing accuracy here. 
- In 2 dimensions, the new implementation is quite a bit slower due to the special case in `mvnun` - perhaps that can be translated separately later, if there are complaints.
- In 3 dimensions, the new implementation seems faster in some cases and slower in others (up to 10x), but on average a bit slower. As the dimensionality increases, the variability is reduced and the performance gap decreases. In 10 dimensions and higher, the new implementation seems to be marginally faster consistently.

If there are concerns about performance in lower dimensionality problems, we can point to this implementation being necessary to close issues. Speaking of which, can you address whether this closes gh-8367 and gh-gh-16142 in the top post?

rkern commented:
Just for reference, the bivariate and trivariate cases can (should?) be ported from the author's latest MATLAB versions (unfortunately now only on the Wayback Machine).

https://web.archive.org/web/20200205123040/http://www.math.wsu.edu/faculty/genz/software/matlab/bvn.m
https://web.archive.org/web/20200205123040/http://www.math.wsu.edu/faculty/genz/software/matlab/tvn.m

There are versions for the [multivariate t distribution, too](https://web.archive.org/web/20200205123719/http://www.math.wsu.edu/faculty/genz/software/matlab/tvt.m).

ev-br committed:
MAINT: stats: cythonize qmvnt

ev-br committed:
MAINT: stats/mvn: address review comments

Plus, fix a small issue in multivariate_normal_frozen: stop
swallowing extra kwargs, e.g. maxpts.

ev-br committed:
MAINT: stats: remove unused _qmvnt_pythran

ev-br commented on scipy/stats/_multivariate.py:
```diff
@@ -906,7 +921,7 @@ def __init__(self, mean=None, cov=1, allow_singular=False, seed=None,
         self.allow_singular = allow_singular or self.cov_object._allow_singular
         if not maxpts:
-            maxpts = 1000000 * self.dim
+            maxpts = 10000 * self.dim
```
In fact, this limit is hit in tests `_vs_univariate_singular`, https://github.com/scipy/scipy/pull/22298/files#diff-a998a313f078eba79aeb6347a65117e0a0c4542a4d778a4cfd398f1737380a71R1090
Bumping it back up makes these tests *slow* again, so the last commit bumps the default back up and limits it in the affected tests.

rkern commented on scipy/stats/tests/test_multivariate.py:
```diff
@@ -810,6 +1038,69 @@ def test_cdf_signs(self):
+                                   # default maxpts is too slow, limit it here
+                                   maxpts=10_000*case.covar.shape[0]
+        )
+        cdf_val = dist.cdf(case.high, rng=rng,)
```
```suggestion
        cdf_val = dist.cdf(case.high, rng=rng)
```

ev-br committed:
TST: stats/mvn: address review comments

mdhaber commented:
Just to see whether it would work, I had ChatGPT translate Genz's bivariate normal implementation to Python, and once I figured out how to transform a problem with general covariance matrix to the required form, it just works. I put it into `_stats_pythran.py`, and execution takes only about twice as long as `_mvn`. (I'm not sure how much it matters, but that's with `np.sum(x * y)` instead of `x @ y` in two places, since `x @ y` gave complaints about BLAS.) Have we discussed a policy for code from generative AI?

rkern commented:
We don't have a generative AI policy yet, I don't think. [We do have this somewhat-inconclusive discussion by the numpy team](https://mail.python.org/archives/list/numpy-discussion@python.org/thread/PME3KHTU54HQNYY5HU5JDDMNGCIWXPOZ/), which of course has significant crossover.

Since your particular task was translating code rather than generating fresh code from a natural language instruction prompt, I'd probably rule that the only copyrightable material in it would be as a derivative of the MATLAB code being translated, so as far as copyright provenance goes, I'd personally be fine with it. I'm somewhat neutral on whether or not we should have a strict "no genAI" policy otherwise.

mdhaber commented:
I'm also not sure. I don't want to start seeing it used to produce code from a natural language prompt, and I don't even want to use it for Array API translations. But I thought it was worth discussing use for translation of a function that a human probably won't need to touch for a long time. I figured that I could translate the Matlab myself, which I would have done very mechanically, or I could let the robot do it (more reliably). In either case, we'd probably treat the code as vendored and review by test rather than line by line against the math, so it seemed like it could be a good application. I'm curious to get @ilayn's opinion with all the Fortran translation effort.

mdhaber commented:
That said, I probably shouldn't side -track this PR. Where are we @ev-br? Is this ready from your perspective?

ev-br commented:
The PR itself is ready from my side. Adding 2d and 3d special cases, and porting mvnun_weighted for gaussian_kde are, I believe, best tackled in follow-up PRs.

The ChatGPT experiment is definitely interesting! I'd agree it's about time we come up with some explicitly articulated stance on genAI contributions; and I totally agree it's best discussed separately. Let's take it to the ML (err, discourse). I can start the discussion over there.

ilayn commented:
I'm quite happy that this is being tackled. So thanks Evgeni and all for this. 

>  I'm curious to get @ilayn's opinion with all the Fortran translation effort.

Just to paste it here before I forget (I'll copy paste it to Discourse once we get going)

If I could, I would have used these ChatGPT/CoPilot/Skynet/"All seeing overlord AI" like there is no tomorrow. But my experience with them so far is that, they are almost good and make just a tiny mistake like `i` becomes `j`, then I spend rest of the week troubleshooting, learning how to use GDB and fail miserably etc. So it seems like we are swapping manual labor of translating with manual labor of troubleshooting, which for me, definitely the latter that is more painful. 

But maybe it is because it is not trained enough on F77 or JurassicPACK code we have. On the other hand, I did get some useful tips if I ask programming tips, say, "show me an example of while loop that uses comma operator as an expression in C17" and it brought some nice results. I did not try but some folks in my team use it for summarizing the code about what it does. 

Other than that, probably @rkern is the most knowledgeable among us what the legal implications might be, however Copilot seems to understand when you say avoid GPL sources for the answer at the citations it brings.

rgommers commented:
> JurassicPACK code

(OT) 😆 for some reason this amuses me on this grey January morning - we should keep this name:)

mdhaber commented:
Looks mostly good, so I'll go ahead and merge now since I know we'll continue to work on this. 

One thing we really need to revisit: as-is, we have `seed` as a parameter of the class initializer, `random_state` as a parameter of `rvs`, and `rng` as a parameter of `cdf`/`logcdf`. Besides that being a bit complicated, `rng` does not follow the behavior specified in SPEC 7, so I'd rather not use that name. Probably the only fix we need to make is to change `rng` to `random_state`, which would also be consistent with the `cdf` method of `multivariate_t`.

The other thing I'd like to study a bit more is the source of the slowdown for dimensionality 4~9. Is it the choice of language (Cython vs Fortran), the algorithm / number of iterations, or is there an obvious bottleneck that we might be able to address? This fixes two big bugs and works toward eliminating some old Fortran, so I think it's worth some performance regression, but it would be good to come up with a response to "bug" reports now and maybe document the rationale for the change in the release notes.

Thanks @ev-br, @rkern!

ev-br commented:
Discourse thread: https://discuss.scientific-python.org/t/a-policy-on-generative-ai-assisted-contributions/1702
